File: plugins/serialization/formats/ebml/swig/python/ebml_reader.c
/************************************************************************************/
/* Copyright (c) 2009 The Department of Arts and Culture,                           */
/* The Government of the Republic of South Africa.                                  */
/*                                                                                  */
/* Contributors:  Meraka Institute, CSIR, South Africa.                             */
/*                                                                                  */
/* Permission is hereby granted, free of charge, to any person obtaining a copy     */
/* of this software and associated documentation files (the "Software"), to deal    */
/* in the Software without restriction, including without limitation the rights     */
/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell        */
/* copies of the Software, and to permit persons to whom the Software is            */
/* furnished to do so, subject to the following conditions:                         */
/* The above copyright notice and this permission notice shall be included in       */
/* all copies or substantial portions of the Software.                              */
/*                                                                                  */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        */
/* THE SOFTWARE.                                                                    */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* AUTHOR  : Richard Carlson, Aby Louw                                              */
/* DATE    : December 2009                                                          */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* C convenience functions for SEbmlRead Python wrapper.                            */
/*                                                                                  */
/*                                                                                  */
/*                                                                                  */
/************************************************************************************/

%define ebml_reader_DOCSTRING
"""
SEbmlRead(path)


EBML data reader class. Create an Ebml reader object with a path to an EBML format file
from which to read. The header of the file is read.

:param path: Full path and file name of the file to read from.
:type path: string

:return: EBML data reader object.
:rtype: SEbmlRead
"""
%enddef

%feature("autodoc", ebml_reader_DOCSTRING) SEbmlRead;


%define ebml_version_DOCSTRING
"""
Get the EBML version of the file.
The version of EBML to which the document conforms to.

:return: EBML version.
:rtype: int
"""
%enddef

%feature("autodoc", ebml_version_DOCSTRING) ebml_version;


%define reader_version_DOCSTRING
"""
Get the EBML reader version of the file.
The minimum EBML version a parser has to support in order to read the document.

:return: EBML reader version.
:rtype: int
"""
%enddef

%feature("autodoc", reader_version_DOCSTRING) reader_version;


%define max_id_width_DOCSTRING
"""
Get the maximum ``ID`` width of the file.
The maximum width of the ``IDs`` used in this document.

:return: maximum ``ID`` width.
:rtype: int
:note: Speect can not read EBML format file with a ``ID`` width greater than 4.
"""
%enddef

%feature("autodoc", max_id_width_DOCSTRING) max_id_width;


%define max_size_width_DOCSTRING
"""
Get the maximum size width of the file.
The maximum width of the size descriptors used in this document.

:return: maximum size width.
:rtype: int
:note: Speect can not read EBML format file with a size width greater than 4.
"""
%enddef

%feature("autodoc", max_size_width_DOCSTRING) max_size_width;


%define doc_type_version_DOCSTRING
"""
Get the document type version of the file.
The version of document type to which the document conforms to.

:return: document type version.
:rtype: int
"""
%enddef

%feature("autodoc", doc_type_version_DOCSTRING) doc_type_version;


%define doc_type_read_version_DOCSTRING
"""
Get the document type reader version of the file.
The minimum ``doc_type_version`` version an interpreter has to
support in order to read the document.

:return: document type reader version.
:rtype: int
"""
%enddef

%feature("autodoc", doc_type_read_version_DOCSTRING) doc_type_reader_version;


%define doc_type_DOCSTRING
"""
doctype()

Get the document type of the file.
An ASCII string that identifies the type of the document.

:return: document type.
:rtype: string
"""
%enddef

%feature("autodoc", doc_type_DOCSTRING) SEbmlRead::doc_type;


%define peek_id_DOCSTRING
"""
peek_id()

Peek the ``ID`` of the next element in the ebml data source.


:return: Ebml ID.
:rtype: int
"""
%enddef

%feature("autodoc", peek_id_DOCSTRING) SEbmlRead::peek_id;


%define skip_element_DOCSTRING
"""
skip_element()

Skip the current ebml element.
"""
%enddef

%feature("autodoc", skip_element_DOCSTRING) SEbmlRead::skip_element;


%define read_uint_DOCSTRING
"""
read_uint()

Read current element as an unsigned integer.

:return: [ID, element]
:rtype: tuple
"""
%enddef

%feature("autodoc", read_uint_DOCSTRING) SEbmlRead::read_uint;


%define read_sint_DOCSTRING
"""
read_sint()

Read current element as a signed integer.

:return: [ID, element]
:rtype: tuple
"""
%enddef

%feature("autodoc", read_sint_DOCSTRING) SEbmlRead::read_sint;


%define read_double_DOCSTRING
"""
read_double()

Read current element as a double.

:return: [ID, element]
:rtype: tuple
"""
%enddef

%feature("autodoc", read_double_DOCSTRING) SEbmlRead::read_double;


%define read_str_DOCSTRING
"""
read_str()

Read current element as a str.

:return: [ID, element]
:rtype: tuple
"""
%enddef

%feature("autodoc", read_str_DOCSTRING) SEbmlRead::read_str;


%define open_container_DOCSTRING
"""
open_container()

Open a container element and return it's ``ID``.

:return: ID
:rtype: int
"""
%enddef

%feature("autodoc", open_container_DOCSTRING) SEbmlRead::open_container;


%define container_exhausted_DOCSTRING
"""
container_exhausted()

Query if all of a currently open container's elements have been read.
Thus, if true then the container is 'exhausted'.

:return: True of False.
:rtype: bool
"""
%enddef

%feature("autodoc", container_exhausted_DOCSTRING) SEbmlRead::container_exhausted;



%apply unsigned int *INOUT { uint32 *id };

%inline
%{
	SObject *sebml_reader_read_object(SEbmlRead *self, uint32 *id, s_erc *error)
	{
		SObject *val;


		val = S_EBMLREAD_CALL(self, read_object)(self, id, error);
		if (*error != S_SUCCESS)
			return NULL;

		return val;
	}
%}


typedef struct
{
	%extend
	{
		const int ebml_version;
		const int reader_version;
		const int max_id_width;
		const int max_size_width;
		const int doc_type_version;
		const int doc_type_reader_version;
	}
} SEbmlRead;


%types(SEbmlRead = SObject);


%extend SEbmlRead
{
	SEbmlRead(const char *path, s_erc *error)
	{
		SDatasource *ds;
		SEbmlRead *ebmlReader;


		/*
		 * create a data source, "rb" = read, binary
		 */
		ds = SFilesourceOpenFile(path, "rb", error);
		if (*error != S_SUCCESS)
			return NULL;

		/*
		 * create ebml reader object
		 */
		ebmlReader = (SEbmlRead*)S_NEW("SEbmlRead", error);
		if (*error != S_SUCCESS)
		{
			S_DELETE(ds, "SEbmlRead()", error);
			return NULL;
		}

		/*
		 * initialize ebml reader object
		 */
		S_EBMLREAD_CALL(ebmlReader, read_init)(&ebmlReader, ds, error);
		if (*error != S_SUCCESS)
		{
			S_DELETE(ds, "SEbmlRead()", error);
			return NULL;
		}

		return ebmlReader;
	}

	const char *doc_type(void)
	{
		return (const char*)$self->header->doctype;
	}

	~SEbmlRead()
	{
		s_erc error;


		S_CLR_ERR(&error);
		S_DELETE($self, "~SEbmlRead()", &error);
	}


	uint32 peek_id(s_erc *error)
	{
		uint32 id;


		id = S_EBMLREAD_CALL($self, peek_id)($self, error);
		if (*error != S_SUCCESS)
			return 0;

		return id;
	}


	void skip_element(s_erc *error)
	{
		S_EBMLREAD_CALL($self, element_skip)($self, error);
	}


	PyObject *read_uint(s_erc *error)
	{
		PyObject *tuple;
		PyObject *object;
		uint32 val;
		uint32 id;


		val = S_EBMLREAD_CALL($self, read_uint)($self, &id, error);
		if (*error != S_SUCCESS)
			return NULL;

		tuple = PyTuple_New(2);
		if (tuple == NULL)
		{
			S_CTX_ERR(error, S_FAILURE,
					  "read_uint",
					  "Call to \"PyTuple_New\" failed");
			return NULL;
		}

		object = PyInt_FromLong((ulong) id);
		PyTuple_SET_ITEM(tuple, 0, object);

		object = PyInt_FromLong((ulong) val);
		PyTuple_SET_ITEM(tuple, 1, object);

		return tuple;
	}


	PyObject *read_sint(s_erc *error)
	{
		PyObject *tuple;
		PyObject *object;
		sint32 val;
		uint32 id;


		val = S_EBMLREAD_CALL($self, read_sint)($self, &id, error);
		if (*error != S_SUCCESS)
			return NULL;

		tuple = PyTuple_New(2);
		if (tuple == NULL)
		{
			S_CTX_ERR(error, S_FAILURE,
					  "read_sint",
					  "Call to \"PyTuple_New\" failed");
			return NULL;
		}

		object = PyInt_FromLong((ulong) id);
		PyTuple_SET_ITEM(tuple, 0, object);

		object = PyInt_FromLong((long) val);
		PyTuple_SET_ITEM(tuple, 1, object);

		return tuple;
	}


	PyObject *read_double(s_erc *error)
	{
		PyObject *tuple;
		PyObject *object;
		double val;
		uint32 id;


		val = S_EBMLREAD_CALL($self, read_double)($self, &id, error);
		if (*error != S_SUCCESS)
			return NULL;

		tuple = PyTuple_New(2);
		if (tuple == NULL)
		{
			S_CTX_ERR(error, S_FAILURE,
					  "read_double",
					  "Call to \"PyTuple_New\" failed");
			return NULL;
		}

		object = PyInt_FromLong((ulong) id);
		PyTuple_SET_ITEM(tuple, 0, object);

		object = PyFloat_FromDouble(val);
		PyTuple_SET_ITEM(tuple, 1, object);

		return tuple;
	}


	PyObject *read_str(s_erc *error)
	{
		PyObject *tuple;
		PyObject *object;
		char *val;
		size_t len;
		uint32 id;


		val = S_EBMLREAD_CALL($self, read_utf8)($self, &id, error);
		if (*error != S_SUCCESS)
			return NULL;

		len = s_strsize(val, error);
		if (*error != S_SUCCESS)
			return NULL;

		tuple = PyTuple_New(2);
		if (tuple == NULL)
		{
			S_CTX_ERR(error, S_FAILURE,
					  "read_str",
					  "Call to \"PyTuple_New\" failed");
			return NULL;
		}

		object = PyInt_FromLong((ulong) id);
		PyTuple_SET_ITEM(tuple, 0, object);

		object = PyString_FromStringAndSize(val, len);
		PyTuple_SET_ITEM(tuple, 1, object);
		S_FREE(val);

		return tuple;
	}


%pythoncode
%{
def read_object(self):
    """
    Read current element as an object.

    :return: [ID, element]
    :rtype: tuple
    :note: The object must have an appropriate *serialization formatter*
           defined and registered for the ``spct_ebml`` format. If not,
           then the object can not be read.
    """

    import speect

    ebml_id = 0
    sobject = sebml_reader_read_object(self, ebml_id)

    # Python takes ownership, TRUE
    pyobject = speect.sobject_2_pobject(sobject, TRUE)

    return [ ebml_id, pobject ]
%}


	uint32 open_container(s_erc *error)
	{
		uint32 id;


		id = S_EBMLREAD_CALL($self, container)($self, error);
		if (*error != S_SUCCESS)
			return 0;

		return id;
	}


	s_bool container_exhausted(s_erc *error)
	{
		s_bool query;


		query = S_EBMLREAD_CALL($self, container_at_end)($self, error);
		if (*error != S_SUCCESS)
			return FALSE;

		return query;
	}
}


%{
	const int SEbmlRead_ebml_version_get(SEbmlRead *self)
	{
		return (int)self->header->ebml_version;
	}

	const int SEbmlRead_reader_version_get(SEbmlRead *self)
	{
		return (int)self->header->ebml_read_version;
	}

	const int SEbmlRead_max_id_width_get(SEbmlRead *self)
	{
		return (int)self->header->max_id_width;
	}

	const int SEbmlRead_max_size_width_get(SEbmlRead *self)
	{
		return (int)self->header->max_size_width;
	}

	const int SEbmlRead_doc_type_version_get(SEbmlRead *self)
	{
		return (int)self->header->doctype_version;
	}

	const int SEbmlRead_doc_type_reader_version_get(SEbmlRead *self)
	{
		return (int)self->header->doctype_read_version;
	}
%}

File: plugins/serialization/formats/ebml/swig/python/ebml_reader.py
######################################################################################
## Copyright (c) 2009 The Department of Arts and Culture,                           ##
## The Government of the Republic of South Africa.                                  ##
##                                                                                  ##
## Contributors:  Meraka Institute, CSIR, South Africa.                             ##
##                                                                                  ##
## Permission is hereby granted, free of charge, to any person obtaining a copy     ##
## of this software and associated documentation files (the "Software"), to deal    ##
## in the Software without restriction, including without limitation the rights     ##
## to use, copy, modify, merge, publish, distribute, sublicense, and#or sell        ##
## copies of the Software, and to permit persons to whom the Software is            ##
## furnished to do so, subject to the following conditions:                         ##
## The above copyright notice and this permission notice shall be included in       ##
## all copies or substantial portions of the Software.                              ##
##                                                                                  ##
## THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       ##
## IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         ##
## FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      ##
## AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           ##
## LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    ##
## OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        ##
## THE SOFTWARE.                                                                    ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## AUTHOR  : Richard Carlson, Aby Louw                                              ##
## DATE    : December 2009                                                          ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## Python class for SEbmlRead plug-in.                                              ##
##                                                                                  ##
##                                                                                  ##
######################################################################################



%pythoncode
%{

class SEbmlRead(speect.SObject):
    """
    An EBML reader class implementation.
    """

    
    def __init__(self, path=None, object=None, owner=False):
        """
        Constructor.
        @param path: Full path and file name of the ebml file to read.
        @type path: str
        @param object: A pointer to a C-type Speect SEbmlRead type
        SObject (default = I{None}).
        @type object: I{PySwigObject SObject*}
        @param owner: I{True} if the Python SEbmlRead is the owner of the
        C-type Speect SObject (and can therefore delete it),
        otherwise I{False} (default = I{False}).
        @type owner: bool
        @return: The newly created EBML reader.
        @rtype: L{SEbmlRead}
        """

        if (path and object) or (not path and not object):
            raise RuntimeError('Either argument \"path\" or '+\
                               '\"object\" must not be \'None\'')

        if path:
            if not isinstance(path, str):
                raise RuntimeError('The path argument must be a \'str\' type')

            ebml_object = py_sebml_reader_new(path)
            if not ebml_object:
                raise RuntimeError('Speect failed to create new \'SEbmlRead\' object')
   
            super(SEbmlRead, self).__init__(ebml_object, True)
        else:
            if not py_sobject_is_type(object, "SEbmlRead"):
                raise TypeError('Input argument \"object\" must be ' +
                                'of type \'C Speect SEbmlRead\'')
            else:
                super(SEbmlRead, self).__init__(object, owner)

        # set header info
        self.ebml_version = py_sebml_reader_get_header_version(ebml_object)
        self.reader_version = py_sebml_reader_get_header_reader_version(ebml_object)
        self.max_id_width = py_sebml_reader_get_header_max_id_width(ebml_object)
        self.max_size_width = py_sebml_reader_get_header_max_size_width(ebml_object)
        self.doc_type = py_sebml_reader_get_header_doctype(ebml_object)
        self.doc_type_version = py_sebml_reader_get_header_doctype_version(ebml_object)
        self.doc_type_reader_version = py_sebml_reader_get_header_doctype_reader_version(ebml_object)


    def peak_id(self):
        """
        Peek the I{id} of the next element.
        @return: The id of the next element.
        @rtype: int
        """

        return py_sebml_reader_peek_id(self._get_speect_object())


    def skip_element(self):
        """
        Skip the next element in the ebml file.
        """

        return py_sebml_reader_element_skip(self._get_speect_object())


    def read_uint(self):
        """
        Read unsigned integer.
        @returns: tuple [ ebml id, unsigned integer ]
        @rtype: tuple [ int, int ]
        """
        
        ebml_id = 0
        val = py_sebml_reader_read_uint(self._get_speect_object(), ebml_id)
        return [ ebml_id, val ]


    def read_sint(self):
        """
        Read signed integer.
        @returns: tuple [ ebml id, signed integer ]
        @rtype: tuple [ int, int ]
        """
        
        ebml_id = 0
        val = py_sebml_reader_read_sint(self._get_speect_object(), ebml_id)
        return [ ebml_id, val ]


    def read_double(self):
        """
        Read double.
        @returns: tuple [ ebml id, double ]
        @rtype: tuple [ int, float ]
        """
        
        ebml_id = 0
        val = py_sebml_reader_read_double(self._get_speect_object(), ebml_id)
        return [ ebml_id, val ]

  
    def read_str(self):
        """
        Read unicode string.
        @returns: tuple [ ebml id, string ]
        @rtype: tuple [ int, unicode ]
        """
        
        ebml_id = 0
        val = py_sebml_reader_read_utf8(self._get_speect_object(), ebml_id)
        return [ ebml_id, unicode(val, 'utf-8') ]


    def read_object(self):
        """
        Read an SObject type instance.
        @returns: tuple [ ebml id, object ]
        @rtype: tuple [ int, L{SObject}/sub-class ]
        """
        
        ebml_id = 0
        object = py_sebml_reader_read_utf8(self._get_speect_object(), ebml_id)
        pyobject = create_py_sobject(item_object, True)
        return [ ebml_id, pyobject ]
    

    def open_container(self):
        """
        Open an EBML container element and return it's id.
        @return: Container id.
        @rtype: int
        """
        
        return py_sebml_reader_read_container(self._get_speect_object())


    def container_exhausted(self):
        """
        Query if the elements in the container have been exhausted.
        @return: I{True} or I{False}.
        @rtype: bool
        """
        
        return py_sebml_reader_container_at_end(self._get_speect_object())
         
  
# register class with Speect Engine
speect.register_class(SEbmlRead)

%}
